### ISSUE TITLE
Incompatible API change in get_player_control() causes server crashes

### ISSUE THREAD

**Issue by erlehmann** [minetest/minetest#11989]
##### Minetest version
Commit 1b2176a426dd987795f20e9042a8d79f958b7b44

##### Summary
In Minetest 5.4.1, calling `get_player_control()` on a mob returned the empty string.

Since commit 5eb45e1ea03c6104f007efec6dd9c351f310193d, calling `get_player_control()` on a mob returns nil.

Several mods contain vessels that can have a human or a mob as a driver. Code like the following now crashes:

```
local ctrl = passenger:get_player_control()
if ctrl.sneak then
    detach_object(passenger, true)
end
```

Furthermore, once a world has crashed, rejoining it near the mob in question immediately crashes it again.

##### Steps to reproduce
1. Start Minetest commit 1b2176a426dd987795f20e9042a8d79f958b7b44 with Mineclonia commit b1b96e3fac6e1fc93356d86132e6b7a65419fd47 in creative mode.
2. Place a boat close to a mob.
3. Enjoy your crash.
4. Try to join the world.
5. Note that the server crashes on joining the world.

##### Error message

```
AsyncErr: Lua: Runtime error from mod '??' in callback luaentity_Step(): ...netest/games/mineclonia/mods/ENTITIES/mcl_boats/init.lua:263: attempt to index local 'ctrl' (a nil value)
stack traceback:
	...netest/games/mineclonia/mods/ENTITIES/mcl_boats/init.lua:263: in function 'on_step_old'
	...mes/mineclonia/mods/ENVIRONMENT/mcl_void_damage/init.lua:17: in function <...mes/mineclonia/mods/ENVIRONMENT/mcl_void_damage/init.lua:16>
```

##### Analysis
I think the incompatible API change was introduced in this PR: https://github.com/minetest/minetest/pull/11924

This means that this commit introduced it: https://github.com/minetest/minetest/commit/5eb45e1ea03c6104f007efec6dd9c351f310193d

The following code was removed from `src/script/lua_api/l_object.cpp`:

```
if (player == nullptr) {
	lua_pushlstring(L, "", 0);
	return 1;
}
```

The following code was added:

```
if (player == nullptr)
	return 0;
```

It seems that this API change was obscured by a refactoring and thus not noticed by reviewers.

##### Affected Games

MineClone2: https://git.minetest.land/MineClone2/MineClone2/issues/1974
MineClone5: https://git.minetest.land/MineClone5/MineClone5/pulls/121
MineClonia: https://git.minetest.land/Mineclonia/Mineclonia/issues/241

**Comment by rubenwardy** [comment]
:-1:  this shouldn't be called for non-players, returning nil is the correct behaviour

**Comment by erlehmann** [comment]
> -1 this shouldn't be called for non-players, returning nil is the correct behaviour. Returning a string is nonsensical

I agree that the original API design makes little sense – but a breaking change in a minor release that crashes mods is *way* worse than having a bad API design. If this bug is not fixed before the release of 5.5, prepare for a bunch of “Minetest crashes every time I enter my world” bug reports.

The correct thing to do here to do is output a deprecation warning … and not crash the server.

I think Minetest should *really* stop using sentimental versioning and start using semantic versioning.

**Comment by erlehmann** [comment]
> The documentation states that player methods are a no-op for entities. Returning a string would be a bug

If that is the case, the documentation lied to all of us.

I have not found documentation on the “nonsensical” use of the function.

But it seems to me that `get_player_control()` just mirrored `get_player_name in returning the empty string.

**Comment by appgurueu** [comment]
The JS way to solve this would be the following:
```lua
debug.setmetatable(nil, {__index = function() return nil end})
```
Put this in builtin and the bug will be fixed :laughing:

**Comment by erlehmann** [comment]
> Side note: Mods can workaround this by altering the ObjRef methodtable, overriding get_player_control() to return the empty string again if it would have returned nil.

Mods can also work around this by changing the code to a check in the style of “if ctrl and ctrl.sneak”.

MineClone5 has already done so – and we are going to change Mineclonia to use that style too.

Again, the issues I see here:

a) Since this is a breaking change introduced without any notice, there was little time to adjust mods.

b) In the real world this results in the game crashing and having a unusable world until a mod is updated.

**Comment by appgurueu** [comment]
> b) In the real world this results in the game crashing and having a unusable world until a mod is updated.

Only if attachments are persisted

**Comment by rubenwardy** [comment]
This is not a breaking change. Defining any change in behaviour of the API as a breaking change will make it very hard to make any changes and fix any bugs in the API. This is an obvious mistake in the API, and the change makes it match the documentation

**Comment by appgurueu** [comment]
> This is not a breaking change. Defining any change in behaviour of the API as a breaking change will make it very hard to make any changes and fix any bugs in the API. This is an obvious mistake in the API, and the change makes it better match the documentation and dev expectations

Nevertheless this is very likely to break many existing mods as @erlehmann has pointed out.

**Comment by erlehmann** [comment]
> Only if attachments are persisted, or automatic reattaching happens (as is the case with MCL)

@appgurueu The detachment is a red herring. The crash happens before anything can be detached.

> This is not a breaking change.

Let me phrase it in a way that you have a harder time redefining phrases such as “breaking change” to whatever:

**The new incompatible API makes existing games and mods crash. Those games and mods did not crash in the past.**

> Defining any change in behaviour of the API as a breaking change will make it very hard to make any changes and fix any bugs in the API.

True. There exists a tried and tested process though: Deprecate this type of thing and throw a warning, for at least one version, before you break previously-working code … which has been done before, as far as I can tell.

> This is an obvious mistake in the API, and the change makes it better match the documentation and dev expectations

I agree that returning the empty string looks like an obvious mistake in the original API.

However, games have been relying on that broken API for about (minimum) eight years now.

You can, of course, claim that changing an API that game and mod devs have relied on for years better matches “dev expectations” because the documentation does not describe the case. However, this does not become magically true because you said it.

**Comment by rubenwardy** [comment]
An alternative would be to return `{}` instead. This would be correct and would avoid these crashes

**Comment by celeron55** [comment]
There's a chance returning {} will break someone's code too, but the likeliness for that is less than 100%. Well, until we hear that it does.

**Comment by erlehmann** [comment]
> An alternative would be to return {} instead. This would be correct and would avoid these crashes

Sure, it could avoid this particular example crash. It would still be a breaking API change and a risk.

IMO, the appropriate time to change APIs in incompatible ways is when they are specced and in major releases.

> The difference is that this is obviously a bug, whereas the moon thing was a matter of convention and could be justified either way

Justifications do not matter. Documentation does not matter. The API existed for years and devs are using it.

No matter how illogical or ugly or undocumented something is, once devs rely on it, Minetest should be careful.

**Comment by celeron55** [comment]
Then, what about get_player_control_bits()? It returned an empty string like this one, and normally returns a number. There's no way to make the API make sense without breaking compatibility.

**Comment by erlehmann** [comment]
@celeron55 good observation. I guess that's another opportunity for a crash waiting to happen. Enjoy!

> There's no way to make the API make sense without breaking compatibility.

I agree. Generally, anyone can make a cleaner API that breaks everything.

It is tedious work and takes long discussions to not break anything.

**Comment by rubenwardy** [comment]
> There's a chance returning {} will break someone's code too, but the likeliness for that is less than 100%. Well, until we hear that it does.

I highly doubt it

**Comment by celeron55** [comment]
{} for get_player_control() and 0 for get_player_control_bits() might just do it, and wouldn't make the API docs look silly.

**Comment by erlehmann** [comment]
Please stop brainstorming incompatible APIs and making unfounded assumptions about how _those_ would be okay.

The only clean solution is to revert the change.

**Comment by celeron55** [comment]
And sounds like that too would work if get_player_control() returned {}? I think it's the way to go.

**Comment by Wuzzy2** [comment]
Thing is, the 5.4.1 docs do not state what happens if you call `get_player_control` on a non-player. The only thing it says is that it is a "no-op" but not what it *returns*.

Therefore, calling `get_player_control` on a non-player 5.4.1 is technically an **undefined return value**. Therefore, it's the fault of any mod that depended on this.

Feel free to add the missing documentation (e.g. that calling these functions on non-players is *forbidden* or that it yields XYZ return value).

Apart from this, no action is needed.

**Comment by erlehmann** [comment]
> Thing is, the 5.4.1 docs do not state what happens if you call get_player_control on a non-player. The only thing it says is that it is a "no-op" but not what it returns.
> 
> Therefore, calling get_player_control on a non-player 5.4.1 is technically an undefined return value. Therefore, it's the fault of any mod that depended on this.

I have been reading opinions similar to this a lot regarding breaking changes.

I think it is not a good idea to handle breakage by focusing on who did something wrong.

As a game developer, I will fix the code to run in Minetest 5.5. Other devs will hopefully do so too.

But think of the users! They are in a bad place if this bug is not fixed. If their worlds are loaded in 5.5.0 and the mods they use are not upgraded, they might not be able to go back to 5.4.0 because the map block compression changed from gzip to zstd … until all mods they use are fixed and they realize that they should upgrade mods.

I had written the above in #minetest-dev already, but since no one else mentioned it, I think it belongs here.

**Comment by corarona** [comment]
> But think of the users! They are in a bad place if this bug is not fixed. If their worlds are loaded in 5.5.0 and the mods they use are not upgraded, they might not be able to go back to 5.4.0 because the map block compression changed from gzip to zstd … until all mods they use are fixed and they realize that they should upgrade mods.

If this is true, or if there is even a chance of it happening - i.e. that you cant load a world you loaded in 5.5 in prior versions - i think this should be communicated *very* clearly.

In the meantime I guess we have to be really careful about 5.5

**Comment by Zughy** [comment]
> The difference is that this is obviously a bug, whereas the moon thing was a matter of convention

No, the moon was a matter of oversight, not convention. I still find the whole story ridiculous, but anyway, have fun with this one

**Comment by sfan5** [comment]
> If this is true, or if there is even a chance of it happening - i.e. that you cant load a world you loaded in 5.5 in prior versions - i think this should be communicated very clearly.

It is true, but this is how automatic migrations in Minetest (and MTG for that matter) have always worked. Major ones are just rarer.

Anyway I think a good fix here is what c55 [proposed](https://github.com/minetest/minetest/issues/11989#issuecomment-1021214246):
> {} for get_player_control() and 0 for get_player_control_bits() might just do it, and wouldn't make the API docs look silly.

**Comment by corarona** [comment]
> It is true, but this is how automatic migrations in Minetest (and MTG for that matter) have always worked. Major ones are just rarer.

There's a joke we like to tell in health: "care could be so nice if it weren't for the patients"

This feels a lot like that except that you're not joking :(

**Comment by sfan5** [comment]
I am intrigued where people get the expectation that Minetest is some magical application that is always compatible even if you don't update your mods for two years, try to load future formats in today's versions or rely on undocumented behaviours.

**Comment by appgurueu** [comment]
> > But think of the users! They are in a bad place if this bug is not fixed. If their worlds are loaded in 5.5.0 and the mods they use are not upgraded, they might not be able to go back to 5.4.0 because the map block compression changed from gzip to zstd … until all mods they use are fixed and they realize that they should upgrade mods.
> 
> If this is true, or if there is even a chance of it happening - i.e. that you cant load a world you loaded in 5.5 in prior versions - i think this should be communicated _very_ clearly.
> 
> In the meantime I guess we have to be really careful about 5.5

Please open another issue for that; users should have an option to decide whether they want their maps aggressively upgraded.

**Comment by erlehmann** [comment]
> I am intrigued where people get the expectation that Minetest is some magical application that is always compatible even if you don't update your mods for two years, try to load future formats in today's versions or rely on undocumented behaviours.

@sfan5 a combination of things influence this:

1. A lot of code from years ago still works – either because it has not been changed in the engine, or because the developers put in backwards-compatible code. This is hugely beneficial … Minecraft, on the other hand, corrupts maps on upgrade AFAIK.

2. Lots of other software APIs broadly follow semantic versioning, thus they increase the major number when . Minetest follows sentimental versioning though (i.e. increase the numbers by how much the core devs think has changed). Maybe it should be emphasized a lot more that Minetest breaks functionality intentionally on almost every official release?

3.  The official documentation is lacking. Often, people have to delve into the code to figure out what works how. Because the semi-official stance from core developers is “if it was not documented for years, we can and will change the API however we please, unless the breakage is too large”, this creates resentment.

4. Projects like the mcl* games are too big for hobbyist developers to be knowledgeful about everything or to be reasonably expected to test everything. At times the code I touch in Mineclonia is years old, written by people who have long since left the project.

5. When stuff is eventually broken or removed, it has usually worked fine for several years. Often it is even obvious to laypeople how to implement it in a somewhat ugly, but backwards-compatible way – so not doing that seems unreasonable to many of the game and mod developers I have chatted with about it.

**Comment by erlehmann** [comment]
> We try our best

**I disagree.**

@sfan5 the Minetest core developers as a group are *not* trying their best *at keeping compatibility across releases* – no one who has worked on a project for years is *that* incompetent. I know that you yourself are capable of fixing complicated bugs and have seen you doing work to ensure backwards compatibility even when mods were doing something obviously wrong, like with the models having too few textures.

IMO an overwhelming amount of backwards compatibility complaints from me and other game & mod developers occur because of structural problems in the development process, all of which could *easily* be improved:

1. As I have said before, Minetest uses *sentimental versioning*, with breaking API changes happening in minor releases as a default. A switch to *semantic versioning* (increase the major version on every breaking change, increase minor version on a backwards compatible change, increase patch level on minor things) would cost next to nothing and make it obvious when anything can be expected to break. The only free software game engine I have seen do much worse on versioning than Minetest is Ren'Py, which I remember to break stuff .

2. Minetest core developers do not always do their research when changing behavior, both as authors and as reviewers. A good example is the now reverted TGA removal: Searching for the string “tga” on ContentDB shows that there exists a mod named `tga_encoder` and going through the list of public servers *right now* shows that it is installed on 9 servers. The PR that did remove it was extremely sloppy work and most likely approved without really going into it (IMO no one can review a 200k line change that fast). Yes, I did intentionally pick the most extreme example I know – because a) it caused a lot of needless conflict b) people asserted falsehoods, like “BMP can be used instead of TGA” or “TGA can not be smaller than PNG” and c) it turned out that *doing the research* changed the outcome of the process (i.e. reintroducing TGA).

3. Minetest developers can be aggressively dismissive of complaints. I understand that some of issues may be bogus and I seem to have filed some myself. Yet, things that seem obvious to a game or mod developer are not always believed. You can probably remember how long I had to argue to convince people of “It has been possible to place nodes out of bounds and it can be a problem in existing code if minetest.find_nodes_in_area() can not find nodes that indeed exist out there.” was finally believed – even though everyone could verify it by emerging the area and placing some nodes out there. Or maybe you remember how you initially reacted to my claims that CMake does not do incremental builds correctly, which are similarly easy to verify (i.e. first build incrementally, then do a clean build & compare results).

4. Trivial requests for reverting breaking changes are rejected with either “we had never documented that, so we can change it freely, update your mod” or “the old API was ugly”. It costs next to nothing to change the attitude that a) someone must be at fault and by default that is the user of the API b) aesthetics are more important than code that actually works.

5. Adding tests hat are not regression tests seems to be frowned upon. I think this is a bad practice. Aggressively adding test cases together with not blocking a merge unless it breaks a test that had previously succeeded is the way to go IMO.

6. PRs and commits often contain several unrelated changes. IMO a refactoring should not also change the behaviour of something, and each thing should se separately revertable if the need arises. Also it is not a big deal to write proper commit messages, yet small changes are often pushed through without those.

7. It seems to be impossible to become a core developer without having contributed some recent features. But for that, the current core developers have to like the features. Ultimately this means that three groups of the community have a hard time to become core developers – despite being able to offer valuable input:

* **Game & mod developers**: They may have little time to contribute features. Often they know much better than core developers when stuff breaks and can give a lot of input on how to prevent such issues. People who put in work like @Wuzzy2 can see their PRs lay around for a long time or be closed due to roadmap issues, while hilarious nonsense like the base 64 encoded PNG in texture modifiers or the PNG encoder that creates a checkerboard texture that optipng can reduce to about 5% of its size is merged. People with their own projects like @appgurueu or me can not easily get into a position where we could approve an important bugfix or where we could veto a useless change.

* **Alternative client developers:** People like @corarona and @EliasFleckenstein03 have experience with CSMs and know how to combat cheating. They could provide valuable advice on accessability and performance issues, which CSM APIs are helpful & help debug issues like coordinate leaks. Yet, any discussion of non-vanilla clients in the forum is forbidden and they are rarely, if ever, asked for advice.

* **Server/Proxy developers**: @HimbeerserverDE and @anon55555 have contributed valuable tooling to the Minetest ecosystem (multiserver, anti-lag proxy, minetest protocol go library). They should have a say in how development continues IMO.

**Comment by erlehmann** [comment]
> but is reasonable that things have to change at some point

@sfan5 for many API changes it would be way less disruptive if they were made opt-in by creating a new function with a numeric suffix that increases for every incompatible API change.

Having functions with names like `minetest.get_player_control2` or `minetest.dynamic_add_media3` may look ugly to some, but:

* It would enable a clean separation of old and new functions, helping deprecation.
* Documentation for an existing function would never go out of date that way.
* Developers could easily search any mod or even ContentDB for usage.

This is only one possible solution among several, but it is obviously easy to implement it.

In general, moving forward while keeping API compatibility is largely a solved problem.

Examples of technology that has moved forward, while keeping backwards compatibility:

* **Shell** scripts written 10 years ago continue to work today.

* **HTML** has added `<audio>` & `<video>` and new input elements.

* **JPEG** images are on par with WebP, through better encoders.

* **HTTP** servers communicate feature improvements with new headers.

* **Python** programs had to be upgraded from 2 to 3, but not otherwise.

* In **CSS** newer values are ignored and older values provide a fallback.

* **XMPP** has been upgraded for OMEMO encryption and improved file uploads.

* **Minetest** supports older clients connecting to newer severs quite well.

* **Email** nowadays supports encryped/signed messages and HTML, through MIME.

It is not surprising that these are things that are widely used by millions of people.

**Comment by Wuzzy2** [comment]
I now changed my position. Revert the return value to whatever "strangeness" it was before, but document it, thus accepting to live with a slightly "ugly" API to prevent something much worse from happening, i.e. breakage.

This was the argument that convinced me to change sides:

> But think of the users! They are in a bad place if this bug is not fixed. If their worlds are loaded in 5.5.0 and the mods they use are not upgraded, they might not be able to go back to 5.4.0 because the map block compression changed from gzip to zstd … until all mods they use are fixed and they realize that they should upgrade mods.

Yeah, I think this can be really ugly ... We can not reasonably expect all relevant mods to be updated. Mods should NOT break after every "minor" release. I no longer think it is fair to blame modders in this case, even if it technically was undefined behavior. The reason is because we're literally talking about crashes here. Also, the "cost" of keeping the "ugly" API like it was before is very low IMHO, it seems the only real argument for changing is "it's ugly now!".

Basically, we have 3 options:

1) Beautiful documented API, but mods/games will crash
2) Ugly documented API, but mods/games will not crash
3) What erlehmann suggested above

> I am intrigued where people get the expectation that Minetest is some magical application that is always compatible even if you don't update your mods for two years, try to load future formats in today's versions or rely on undocumented behaviours.
We try our best but is reasonable that things have to change at some point

If it is official policy to not care about compability, then Minetest should at least be honest about it and literally make a major release every release, i.e. 6.0.0, 7.0.0, 8.0.0, etc., in accordance to SemVer, unless there really is no compability break for once. :P

**Comment by Wuzzy2** [comment]
To be clear: I still think modders should in generally NOT rely on documented or obviously buggy behavior and then act surprised when it breaks in future. But in **this particular case**, I think the price of just keeping everything the way it was before, but documented, is very low, compared to the price of intentionally breaking mods that "broke the rules". The whole discussion is basically only about which return value is the "correct" one …

So the question this really boils down to, is this: What is more important to us: Prevent breakage, or have an "beautiful" API (whatever that means)?

**Comment by Wuzzy2** [comment]
Correction: Of course, I meand *undocumented* behavior.

**Comment by sfan5** [comment]
> If it is official policy to not care about compability, then Minetest should at least be honest about it and literally make a major release every release, i.e. 6.0.0, 7.0.0, 8.0.0, etc., in accordance to SemVer, unless there really is no compability break for once. :P

I can't tell if this is a serious suggestion but I'll try to explain again:
<table>
<tr><td>
Minetest 7.0 is incompatible with 6.0 except for: <br/>
<ul><li>Old maps can be loaded seamlessly
</li><li>Old clients can connect to new servers and vice-versa
</li><li>Configuration and graphic settings work like before
</li><li>Texture packs continue to work like before
</li><li>The vast majority of games or mods do not need any changes
</li><li>The same model, texture and sound formats as before are accepted
</li><li>System requirements (OS/Platform support) have not been increased
</li></ul>
</td><td>
As usual Minetest 6.1 is backwards compatible to 6.0, but take note of the following changes:<br/>
<ul><li>The <code>minetest.foo_bar_baz()</code> method now requires N+1 parameters, the previous behaviour was deprecated in the last release.</li></ul>
</td></tr>
</table>

Consider these two fictional release notes that describe the <b>exact same changes</b>, I think you can tell which is more practical.

In particular since "breaking change" is defined as "something which existed changed in a way it wasn't before" literally every release would be a major (breaking) release. This would also means since it's already <i>allowed</i> by semver then making bigger and more breaking changes would be perfectly fine. But at least semver is now followed by the letter, right?


> The whole discussion is basically only about which return value is the "correct" one …

Well this discussion is already done, we decided to not change the API in incompatible ways <i>even though</i> the behavior was undocumented.
What's left now is just certain people in this issue claiming that nothing must change under any circumstances.

**Comment by erlehmann** [comment]
@sfan5 your fictional release notes are a strawman. In reality, it would look like this:

> Minetest 7.0 is mostly backwards compatible to 6.0. Game and mod developers are advised that things may break:
>
>    The minetest.foo_bar_baz() method now requires N+1 parameters, the previous behaviour was deprecated in the last release.
>    Loading Quake2 models is no longer supported. Minetest 6.0 already warned when loading them and we found no mods in the forums, on contentdb, on GitHub, GitLab or MeseHub that made use of this functionality.
>    Minetest 7.0 requires GCC 18 due to usage of some C++31 features and can not be built with previous versions anymore.

**Comment by Wuzzy2** [comment]
The fictional release notes don't make sense and indeed look like a strawman. Deprecations are fine and are not breaking changes. I also don't consider literally any change as "breaking". Adding a new optional function parameter is also not breaking. Changing a function in a way that breaks existing mods and **causes crashes** *is* a breaking change.

I agree Minetest is already trying hard to not break with compability. Which is good. But it isn't strict enough. There are always a few exceptions to the rule, and I just want to get rid of the annoying exceptions. The past versions had following breaking changes:

* The change of `minetest.PLAYER_MAX_BREATH_DEFAULT`. Mods were forced to change, no compability layer, no fault of modders. Sadly, I *still* see servers with a broken breathbar because of this change. :-(
* Object crosshair. Mods were forced to add an image to keep the old graphics, no fault of modders.
* Ores break when `noise_params` was missing (OK, in this case you might blame modders for this)

I argue that these changes *could* have been done in a non-breaking manner, but it was refused to do so.

I think when Minetest has a minor update, mods should still behave like before, with no major or breaking changes (within a reasonable frame). This means, if the modder does nothing, the mod still works as expected. **Thre should be no surprises**. But when the modder has to actively do something only for the mod to go back to the previous behavior, that's what I have a problem with. The sad part about this is that a mod that worked 100% perfect and bug-free in one version will slowly detoriate over time only because of new Minetest versions.

Note I said "within a reasonable frame", this means that not every change needs to be considered to be a breaking change. Stuff like tweaking the lighting curve, or sky gradient is probably fine.

And as for "ugly code/API cleanup, I guess the best way forward is to just deprecate and when a major release is upcoming, delete the legecy trash that has piled up over the years. ;-)

Note I did not add "bumpmap texture removal" to the list. Why? Because this was a feature that was very bad and broken to begin with, and even then, future versions didn't crash. I think it is also fine to remove very broken features provided the future version doesn't crash outright.

**Comment by sfan5** [comment]
> The fictional release notes don't make sense and indeed look like a strawman.

Why? I admit my argument takes a quick detour but it is logical to me.
If you do major releases you set the users' expectations for breaking changes so you'd want some detailed document (the release notes) that confirms that most things are in fact compatible. To minimize the difference between expectations and reality it'd make more sense to do minor releases (= claim that it's compatible) and then have the release notes explain what changed in an _incompatible_ way. The latter is what we do now.

> I think when Minetest has a minor update, mods should still behave like before, with no major or breaking changes (within a reasonable frame). This means, if the modder does nothing, the mod still works as expected. **There should be no surprises.**

This doesn't quite make sense together with "Deprecations are fine and are not breaking changes.".
Do you only want deprecated behavior to be removed on a major release? That's certainly possible but comes at the expense of being able to add new features. Unless of course you make *every* release a major one in which case see above.

**Comment by Wuzzy2** [comment]
(Let's move the general "breaking change" discussion to #12004)

If I would write a PR that reverts the `get_player_control` return value to whatever we had in 5.4.0, but retroactively added the missing documentation on the return value, would it get accepted? It's a trivial change, I guess.
Personally, I don't really care if the return value is 0, false, nil or empty string as long it doesn't break things. Yes, it is *slightly* weird that an empty string was returned but I don't really mind that much keeping it if it keeps compability.

Again: We're talking about **crashes** here, this is not minor. Can we agree on that it is bad when an upgrade to a minor version causes crashes?

If this PR would be about removing a feature that was completely broken beyond repair, then I would understand. But this is not the case.

If you are still against this, I have a different proposal: Add documentation that the return value of all `player` functions is `nil` when called on a non-player. Probably should also *test* all those functions, just in case there is another weird function. Guess what, this is what I will do now …

**Comment by sfan5** [comment]
Who are you talking to then?
I neither am nor was I against fixing this issue. All my recent replies were in context of the "breaking changes" discussion.

I guess we're done here regardless.

**Comment by appgurueu** [comment]
> I guess we're done here regardless.

Yes! This is probably the most discussion I've ever seen on a closed Minetest issue :laughing:, and most of it belongs in #12004.
